_posts/2019-08-18-Securing-Web-Applications-With-Keycloak.md

```diff
@@ -0,0 +1,608 @@
+          sendAccessToken: true
+      }
+  })
+```
+
+Notice how we configure `localhost:9090/api` to be the only URL where we will send our access token to.  
+
+
+Next up, let's add log in and log out buttons in `app.component.html`.
+
+```html
+<button class="btn btn-default" (click)="login()">
+  Login
+</button>
+<button class="btn btn-default" (click)="logoff()">
+  Logout
+</button>
+```
+
+Now all there is left to do is to configure the `OAuthService`. We do this in `app.component.ts`:
+
+```typescript
+import { Component } from '@angular/core';
+import { OAuthService, NullValidationHandler, AuthConfig } from 'angular-oauth2-oidc';
+import { JwksValidationHandler } from 'angular-oauth2-oidc';
+
+@Component({
+  selector: 'app-root',
+  templateUrl: './app.component.html',
+  styleUrls: ['./app.component.css']
+})
+export class AppComponent {
+  title = 'Tour of Heroes';
+  
+  constructor(private oauthService: OAuthService) {
+    this.configure();
+  }
+
+  authConfig: AuthConfig = {
+    issuer: 'http://localhost:8080/auth/realms/heroes',
+    redirectUri: window.location.origin + "/heroes",
+    clientId: 'spa-heroes',
+    scope: 'openid profile email offline_access heroes',
+    responseType: 'code',
+    // at_hash is not present in JWT token
+    disableAtHashCheck: true,
+    showDebugInformation: true
+  }
+  
+  public login() {
+    this.oauthService.initLoginFlow();
+  }
+  
+  public logoff() {
+    this.oauthService.logOut();
+  }
+  
+  private configure() {
+    this.oauthService.configure(this.authConfig);
+    // this.oauthService.tokenValidationHandler = new JwksValidationHandler();
+    this.oauthService.tokenValidationHandler = new NullValidationHandler();
+    
+    this.oauthService.loadDiscoveryDocumentAndTryLogin();
+  }
+}
+```
+
+Make sure that the name of the realm (heroes) and the client id (spa-heroes) correspond to the ones you defined in Keycloak. 
+Remember how we required the heroes scope to be present in our back end? The scope property is how we fix this. 
+If we omit `heroes` from the scope list, we will be getting a 403 response from our resource server.
+
+> Keycloak is not returning the at_hash claim in the access token. For this reason, the client library would crash while parsing it.
+This is why we disable it in the config but also use the `NullValidationHandler` as it would also make the application crash.
+
+We can now try to log in and if all went well, we should see our heroes appear again when browsing to `/dashboard` or `/heroes`.
+
+> Don't worry if you do not immediatly see the heroes appear. This is because we load the `/heroes` page before our code was exchanged for an access token.
+This results in the first `/heroes` call getting a 401 response. You can create a new endpoint and use it as the redirectUri to get rid of this problem.
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/result.gif" width="70%">
```

Comment:
Add an image with eg. `<img class="image fit" src="{{ '/img/devoxx-ma-2018/speakers-dinner.jpg' | prepend: site.baseurl }}" alt="Speakers Dinner" width="60%" />`, otherwise this won't scale good.